src/main/java/org/swisspush/reststorage/s3/S3FileSystemStorage.java

```diff
@@ -0,0 +1,386 @@
+            } else {
+                try {
+                    Files.createDirectory(fullPath.getParent());
+                    putFile(handler, fullPath);
+                } catch (IOException e) {
+                    log.error("Failed to create directory '{}'", path);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void put(String path, String etag, boolean merge, long expire, String lockOwner, LockMode lockMode, long lockExpire, boolean storeCompressed, Handler<Resource> handler) {
+        if (storeCompressed) {
+            log.warn("PUT with storeCompressed option is not yet implemented in file system storage. Ignoring storeCompressed option value");
+        }
+        put(path, etag, merge, expire, "", LockMode.SILENT, 0, handler);
+    }
+
+    private void putFile(final Handler<Resource> handler, final Path fullPath) {
+        try {
+            // DON"T close it in finally block, async code
+            final OutputStream outputStream = Files.newOutputStream(fullPath);
+            final DocumentResource d = new DocumentResource();
+
+            FileWriteStream fileWriteStream = new FileWriteStream(outputStream);
+            d.writeStream = fileWriteStream;
+
+            final Runnable cleanUp = () -> {
+
+                try {
+                    outputStream.close();
+                } catch (IOException e) {
+                    log.error("Failed to close output stream:", e);
+                }
+                fileWriteStream.end();
+            };
+
+            d.closeHandler = event -> {
+                cleanUp.run();
+                d.endHandler.handle(null);
+            };
+            d.addErrorHandler(err -> {
+                log.error("Put file failed:", err);
+                cleanUp.run();
+            });
+            handler.handle(d);
+        } catch (IOException e) {
+            log.error("Failed to put file:", e);
+        }
+    }
+
+    @Override
+    public void delete(String path, String lockOwner, LockMode lockMode, long lockExpire, boolean confirmCollectionDelete,
+                       boolean deleteRecursive, final Handler<Resource> handler) {
+
+        // at this stage I don't know what of this path point to
+        final Path fullFilePath = canonicalize(path, false);
+        final Path fullDirPath = canonicalize(path, true);
+
+        Resource resource = new Resource();
+        boolean deleteRecursiveInFileSystem = true;
+        if (confirmCollectionDelete && !deleteRecursive) {
+            deleteRecursiveInFileSystem = false;
+        }
+        boolean finalDeleteRecursiveInFileSystem = deleteRecursiveInFileSystem;
+
+        // try to check as a file
+        boolean exists = Files.exists(fullFilePath, LinkOption.NOFOLLOW_LINKS);
+        Path fullPath = fullFilePath;
+
+        if (!exists) {
+            // not exists, maybe is a directory
+            exists = Files.exists(fullDirPath, LinkOption.NOFOLLOW_LINKS);
```

Comment:
The same comment as before regarding the remarks on the directories.

Reply:
this still go though java nio library, so need create the directory before access it